File: app/leetcode/No75.sortColors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function (nums) {

    if (nums.length <= 1) return

    // 思路1，计数排序
    // const counter = new Array(3).fill(0)
    // for (const num of nums) {
    //     counter[num]++
    // }
    // let j = 0
    // for (let i = 0; i < counter.length; i++) {
    //     let count = counter[i]
    //     while (count--) {
    //         nums[j++] = i
    //     }
    // }

    function swap(i, j) {
        [nums[i], nums[j]] = [nums[j], nums[i]]
    }

    // 荷兰国旗问题
    // 思路2：两次遍历
    // const n = nums.length
    // let ptr = 0
    // // 将0放到正确的位置
    // for (let i = 0; i < n; i++) {
    //     if (nums[i] === 0) {
    //         swap(i, ptr++)
    //     }
    // }
    // // 将1放到正确的位置，剩下的2自然而然位于正确的位置
    // for (let i = ptr; i < n; i++) {
    //     if (nums[i] === 1) {
    //         swap(i, ptr++)
    //     }
    // }

    // 思路3：双指针
    // let p0 = 0, p2 = nums.length - 1, cur = 0
    // while (cur <= p2) {
    //     const num = nums[cur]
    //     if (num === 0) {
    //         swap(cur, p0)
    //         cur++
    //         p0++
    //     } else if (num === 2) {
    //         swap(cur, p2)
    //         // 注意：！！！
    //         // cur++
    //         p2--
    //     } else {
    //         cur++
    //     }
    // }

    // 思路4：用3个指针p0,p1,p2记录0,1,2下标的位置，初始值都是-1
    // 如果nums[i] = 0,则p2向后移动1位，p1向后移动1位，将p0位置的值设置为0，并自增
    // 如果nums[i] = 1，则会对p2有影响，p2向后移动1位
    // 如果nums[i] = 2，则直接插入p2
    let p0 = -1, p1 = -1, p2 = -1
    for (let i = 0; i < nums.length; i++) {
        const num = nums[i]
        if (num === 0) {
            nums[++p2] = 2
            nums[++p1] = 1

            nums[++p0] = 0
        } else if (num === 1) {
            nums[++p2] = 2

            nums[++p1] = 1
        } else {
            nums[++p2] = 2
        }
    }
};

const nums = [2, 0, 2, 1, 1, 0]
sortColors(nums)
console.log(nums);

File: app/leetcode/No75.sortColors.py
from typing import List

class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # 计数排序
        # counter[0] 记录0的个数
        # counter[1] 记录1的个数
        # counter[2] 记录2的个数
        counter = [0] * 3
        for num in nums:
            counter[num] += 1
        i = 0
        for j in range(3):
            for _ in range(counter[j]):
                nums[i] = j
                i += 1

    def sortColors2(self, nums: List[int]) -> None:
        """
        三指针：
        i 元素为 0 的右边界，初始为 -1
        j 元素为 2 的左边界，初始为 n
        k 当前遍历元素，初始为 0

        当 k < j 时：
        1. nums[k] == 0, nums[k] 和 nums[i+1] 交换，i += 1，k += 1
        2. nums[k] = 2, nums[k] 和 nums[j-1] 交换，j -= 1
        3. nums[k] == 1, k += 1
        
        遍历结束后数组元素分为了 3 个部分 [0, i], [i+1,j-1], [j,n-1] 3 个部分的元素恰好都是 0，1，2
        """
        i, j, k = -1, len(nums), 0
        while k < j:
            if nums[k] == 0:
                i += 1
                nums[i], nums[k] = nums[k], nums[i]
                k += 1
            elif nums[k] == 2:
                j -= 1
                nums[j], nums[k] = nums[k], nums[j]
            else:
                k += 1

nums = [2,0,2,1,1,0]
Solution().sortColors(nums) 
print(nums)